File: javascript/lang/statefulness/module-level-mutable-state.js
// ruleid: module-level-mutable-state
const sessions = new Map();

// ruleid: module-level-mutable-state
let pendingJobs = [];

// ruleid: module-level-mutable-state
let featureOverrides = {};

// ruleid: module-level-mutable-state
const seenIds = new Set();

// ok: module-level-mutable-state
const DEFAULT_HEADERS = { 'content-type': 'application/json' };

// ok: module-level-mutable-state
const RETRY_LIMIT = 3;

function handleRequest(req) {
  // ok: module-level-mutable-state
  const scratch = new Map();
  // ok: module-level-mutable-state
  let collected = [];
  collected.push(req.body);
  return { scratch, collected };
}

const transform = (rows) => {
  // ok: module-level-mutable-state
  const index = new Map();
  for (const row of rows) index.set(row.id, row);
  return index;
};

class Aggregator {
  constructor() {
    // ok: module-level-mutable-state
    this.items = [];
  }

  reduce() {
    // ok: module-level-mutable-state
    const buckets = new Map();
    return buckets;
  }
}

module.exports = { sessions, pendingJobs, handleRequest, transform, Aggregator };

File: python/lang/statefulness/module-level-mutable-state.py
import collections
import queue
from collections import defaultdict, deque
from typing import TYPE_CHECKING

# ruleid: module-level-mutable-state
sessions = {}

# ruleid: module-level-mutable-state
_pending_jobs = []

# ruleid: module-level-mutable-state
seen_ids = set()

# ruleid: module-level-mutable-state
counters = collections.defaultdict(int)

# ruleid: module-level-mutable-state
recent = deque(maxlen=100)

# ruleid: module-level-mutable-state
work = queue.Queue()

# ruleid: module-level-mutable-state
lookup = defaultdict(list)

# ok: module-level-mutable-state
DEFAULT_HEADERS = {}

# ok: module-level-mutable-state
ALLOWED_HOSTS = []

if TYPE_CHECKING:
    # ok: module-level-mutable-state
    annotations = {}


def handler(event):
    # ok: module-level-mutable-state
    local_buffer = []
    # ok: module-level-mutable-state
    local_index = {}
    local_buffer.append(event)
    return local_index


async def async_handler(event):
    # ok: module-level-mutable-state
    scratch = {}
    return scratch


class Aggregator:
    # ok: module-level-mutable-state
    registry = {}

    def __init__(self):
        # ok: module-level-mutable-state
        self.items = []
